addons/mail/static/src/new/public/discuss_public.js

```diff
@@ -14,14 +14,38 @@ export class DiscussPublic extends Component {
 
     setup() {
         this.messaging = useMessaging();
+        this.threadService = useState(useService("mail.thread"));
         this.store = useStore();
-        this.state = useState({ welcome: true });
+        this.state = useState({
+            welcome: this.props.data.discussPublicViewData.shouldDisplayWelcomeViewInitially,
+        });
+        useEffect(() => {
+            !this.state.welcome &&
+                this.threadService.fetchChannelMembers(thread) &&
+                this.switchToThreadView();
+        });
         this.threadService = useService("mail.thread");
-        const thread = Thread.insert(this.store, {
+        this.isChannelTokenSecret = true;
+        const thread = this.thread;
+        this.threadService.setDiscussThread(thread);
+    }
+
+    get thread() {
+        return Thread.insert(this.store, {
             id: this.props.data.channelData.id,
             model: "mail.channel",
             type: this.props.data.channelData.channel.channel_type,
+            serverData: { uuid: this.props.data.channelData.uuid },
         });
-        this.threadService.setDiscussThread(thread);
+    }
+
+    switchToThreadView() {
+        if (this.isChannelTokenSecret) {
+            window.history.replaceState(
+                window.history.state,
+                null,
+                `/discuss/channel/${this.thread.id}${window.location.search}`
+            );
```

Comment:
Would keep comment on why this is needed.
Would have been best to make a test, so it's clear. But it doesn't look easy to make it testable, so at least a comment would help.